fix for #172

tokejepsen opened the pull request:
Fix for #172

mottosso commented:
Nice catch.

It looks like this runs, even when the icon is already visible. Since it runs once per result, and is a relatively complex loop, how about something like this before it?

```python
if item.actionsIcon:
  continue
```

That way, it only happens once.

Also, let's give `actionsIcon` a better name. It sounds like the icon for actions, when it acts like an indicator about whether it should be visible.

Does `actionsIconVisible` better state its intent?

Also, let's get a test in there as well. ;)

tokejepsen commented:
> Also, let's get a test in there as well. ;)

Questions is what I'm testing for?
Technically everything was working correctly, that problem was that the actions list got emptied.

mottosso commented:
Test for the expected behaviour.

The idea is that, whenever this breaks again, it will be caught by a test.

tokejepsen commented:
There we are:)

mottosso commented:
Almost got it!

This test would pass even if it didn't fail and there was an automatically added action, like a default action.

We need to make sure it doesn't pass when the problem is there, and try and limit the range of the test to the tiny portion it concerns.

Indexing into the `.plugins` list is fragile too; if Pyblish or a plug-in had decided to rearrange the registered plug-ins after they registered, this test would also fail; even though the thing it's testing should have been unaffected.

Using `.plugins["MyPlugin"]` is safer.

tokejepsen commented:
I thought the tests were isolated from other tests/plugins?

But yeah, I'll have a look at getting the plugins explicitly.

mottosso commented:
> I thought the tests were isolated from other tests/plugins?

How do you mean? If you're registering a plug-in in the same test, then it's safe to refer to explicitly. Just make sure you de-register them when you're done, to not cause side-effects for other tests. If you're unsure about how to do that that, use a unique name for the plug-in.

tokejepsen commented:
> Just make sure you de-register them when you're done, to not cause side-effects for other tests.

Alright, thought for some reason that the decorator took care of any cross test contamination, so the tests were isolated from each other.

mottosso commented:
Ah, yes you are right, Didn't notice you were using it!

tokejepsen commented:
I've updated the test to explicitly getting the plugins by name.
Also changed the name of the test, as it was the same as the one I copied from :worried:

BigRoy commented:
Wanted to say thanks again for jumping onto this! :)

By the way, looking at the test. Should it maybe also include a case where it doesn't have any actions?

mottosso commented:
Yes, thanks a lot!

> Should it maybe also include a case where it doesn't have any actions?

I think so too. I'm having a look at a few more tests for this now.

mottosso commented on tests/test_control.py:
```diff
@@ -402,28 +402,72 @@ def process(self, instance):
+
+    validate(c)
+
+    assert c.item_model.plugins["ValidateFailA"].actions
```
I've had a closer look at this, and I'm not sure this is right. The plug-in should always contain it's actions, it's only when asking for them that the filtering takes place.

- https://github.com/pyblish/pyblish-qml/blob/87cf2c8dfe500e5dd41d9d6c8e0005a45f3445ff/pyblish_qml/control.py#L302

tokejepsen replied:
The problem wasn't in the controller, but in the model; https://github.com/pyblish/pyblish-qml/blob/87cf2c8dfe500e5dd41d9d6c8e0005a45f3445ff/pyblish_qml/models.py#L431

The problem was that the it didn't maintain its original actions, because the model emptied the actions list by not copying the list.

mottosso replied:
But was the test correct as was merged just now? It would have passed regardless of the `on` state I think.

tokejepsen replied:
> But was the test correct as was merged just now?

It failed with previous code, and passed with the fixed code. Or maybe I don't understand what you are asking?

mottosso replied:
It passed because it would have passed regardless of a plug-in failing.

That list contains all actions of the plug-in, not the currently available actions in the GUI, that's what the [`getPluginActions()`](https://github.com/pyblish/pyblish-qml/blob/29c24ddbafaf2b6edc365ee3027596c580dd0e08/pyblish_qml/control.py#L302) method is for. That will take into account whether anything has failed or not.

I modified your test, and added a plug-in that did not fail. Have a look and see if it makes sense.

- https://github.com/pyblish/pyblish-qml/blob/master/tests/test_control.py#L487

tokejepsen replied:
Cool, yeah, looks good to me.
